From: Raphael Manfredi <Raphael_Manfredi@pobox.com>
Date: Tue, 26 Jan 2021 08:55:43 +0100
Applied-Upstream: https://github.com/logrotate/logrotate/commit/88b2de7bc6d57134e118c21655ae05d6461dc8ae
Subject: Only attempt to set user/group if running as root.

There is no need to attempt to change the user/group of files if we are
not running as root.  We can leave the permissions to what the user
running logrotate would get if doing the same operation manually by
copying the file!

Closes: https://github.com/logrotate/logrotate/pull/372
---
 logrotate.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/logrotate.c b/logrotate.c
index 0a778bf..88003db 100644
--- a/logrotate.c
+++ b/logrotate.c
@@ -582,8 +582,12 @@ static int createOutputFile(const char *fileName, int flags, const struct stat *
         return -1;
     }
 
-    if ((sb_create.st_uid != sb->st_uid || sb_create.st_gid != sb->st_gid) &&
-            fchown(fd, sb->st_uid, sb->st_gid)) {
+    /* Only attempt to set user/group if running as root */
+    if (
+        ROOT_UID == geteuid() &&
+        (sb_create.st_uid != sb->st_uid || sb_create.st_gid != sb->st_gid) &&
+        fchown(fd, sb->st_uid, sb->st_gid)
+    ) {
         message(MESS_ERROR, "error setting owner of %s to uid %u and gid %u: %s\n",
                 fileName, (unsigned) sb->st_uid, (unsigned) sb->st_gid, strerror(errno));
         close(fd);
